### PR TITLE
[MOB-17072] Adopt changes for Core 2.0 part 1

### PR DESCRIPTION
**Pull request by kevinlind**
## Description

Part 1 of adoption of changes for Core 2.0

- Updates .gradle files to use Core 2.0 from jitpack
- Sets extension version to 2.0.0

- Updates Edge public APIs for Core 2.0
- Use new shared state api in EdgeExtension, EdgeState, EdgeHitProcessor
- Implements EdgeExtension.onRegistered and registers listeners. Removes older Listener class
- Implements EdgeExtension.readyForEvent which calls EdgeState.bootupIfNeeded and checks shared state dependencies for specific events.
- Updates EdgeState and EdgeProperties for Core 2.0 changes.
- Switch to new Log class only for the class touched by the changes above.

- Set permission for `.githooks/pre-commit` for execution

Out of scope for this PR:
- Switch to use public EventType and EventSource in all places
- Switch to use new Log class in all places
- Updates to hit queue, EdgeNetworkService, and NetworkResponseHandler except to get code to compile
- Unit and functional tests do not compile or run

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java [hunk 1]
```diff
@@ -302,13 +304,22 @@ private Map<String, String> getRequestHeaders() {
...
+			return requestHeaders;
+		}
+
+		final String assuranceIntegrationId = DataReader.optString(
+			assuranceStateResult.getValue(),
+			EdgeConstants.SharedState.Assurance.INTEGRATION_ID,
+			null
+		);
+
+		if (!StringUtils.isNullOrEmpty(assuranceIntegrationId)) {
```
Comment (kevinlind) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeHitProcessor.java:322]: I've only updated EdgeHitProcessor to fix uses of getSharedState and to use the new DataReader in this `getRequestHeader` function. The reset of this class will be updated in another PR.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeNetworkService.java [hunk 2]
```diff
@@ -339,7 +342,7 @@ public void call(final HttpConnecting connection) {
 			countDownLatch.await();
 			return httpConnecting[0];
 		} catch (final InterruptedException | IllegalArgumentException e) {
-			Log.warning(LOG_TAG, "Connection failure for url (%s), error: (%s)", url, e);
+			Log.warning(LOG_TAG, LOG_SOURCE, "Connection failure for url (%s), error: (%s)", url, e);
```
Comment (kevinlind) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeNetworkService.java:345]: I've only updated the EdgeNetworkService to fix this one instance of `Log` to get the code to compile. The reset of the class will be updated in another PR.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeEndpoint.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 package com.adobe.marketing.mobile;
 
+import com.adobe.marketing.mobile.util.StringUtils;
```
Comment (kevinlind) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeEndpoint.java:14]: This line was the only needed change. Please review entire class.

File: code/edge/src/main/java/com/adobe/marketing/mobile/NetworkResponseHandler.java [hunk 4]
```diff
@@ -31,6 +33,8 @@
  */
 class NetworkResponseHandler {
 
+	private static final String LOG_SOURCE = "NetworkResponseHandler";
```
Comment (kevinlind) [inline code/edge/src/main/java/com/adobe/marketing/mobile/NetworkResponseHandler.java:36]: I've only updated NetworkResponseHandler to fix the cases of `Log` which were failing to compile. The rest of this class will be updated in another PR.

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeStateTests.java [hunk 5]
```diff
@@ -44,7 +44,7 @@ public void setUp() {
 		mockSharedStateCallback =
 			new EdgeSharedStateCallback() {
 				@Override
-				public Map<String, Object> getSharedState(final String stateOwner, final Event event) {
+				public SharedStateResult getSharedState(final String stateOwner, final Event event) {
```
Comment (kevinlind) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeStateTests.java:47]: The changes in EdgeStateTests were done automatically when I made the API changes. The test classes will be updated in another PR.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeConstants.java [hunk 6]
```diff
@@ -14,7 +14,7 @@
 final class EdgeConstants {
 
 	static final String EDGE_DATA_STORAGE = "EdgeDataStorage";
-	static final String EXTENSION_VERSION = "1.4.0";
+	static final String EXTENSION_VERSION = "2.0.0";
```
Comment (kevinlind) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeConstants.java:17]: As this PR is just part 1 of the changes to adopt Core 2.0, I'm not able to remove any unneeded constants, such as the EdgeType and EdgeSource constants.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java [hunk 7]
```diff
@@ -81,279 +87,137 @@ protected EdgeExtension(final ExtensionApi extensionApi) {
...
-							LoggingMode.WARNING,
-							LOG_TAG,
-							"EdgeExtension - There was an error registering Extension Listener for shared state events: " +
-							extensionError.getErrorName()
-						);
-					}
-				}
-			);
+	@Override
+	protected String getFriendlyName() {
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:115]: It's new for Core 2.0. Internal extensions already had a friendly name but 3rd party extensions didn't, so in the EventHub shared state their friendly name was the same as the extension name. Core 2.0 added this function so all the extensions now have friendly names.

File: code/edge/src/main/java/com/adobe/marketing/mobile/Edge.java [hunk 8]
```diff
@@ -13,12 +13,21 @@
...
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
 
 @SuppressWarnings("unused")
 public class Edge {
 
+	public static final Class<? extends Extension> EXTENSION = EdgeExtension.class;
+
+	private static final String LOG_SOURCE = "Edge";
+	private static final long CALLBACK_TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(1);
```
Comment (cacheung) [inline code/edge/src/main/java/com/adobe/marketing/mobile/Edge.java:29]: Should this be 5000?

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeState.java [hunk 9]
```diff
@@ -55,11 +57,21 @@ class EdgeState {
...
-	void bootupIfNeeded(final Map<String, Object> registeredExtensionsWithHub) {
+	boolean bootupIfNeeded() {
 		if (hasBooted) {
-			return;
+			return true;
+		}
+
+		// Get Hub's shared state needed to check if Identity and Consent are registered
+		// If not set, return false and attempt bootup at next event
+		final SharedStateResult eventHubStateResult = sharedStateCallback.getSharedState(
```
Comment (prudrabhat) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeState.java:69]: Does the actual check on whether Consent and Identity are registered happen elsewhere?

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java [hunk 10]
```diff
@@ -363,29 +227,28 @@ void handleSharedStateUpdate(final Event event) {
...
 
 	/**
 	 * Handles the Consent Update event
 	 *
 	 * @param event current event to process; the event and the event data should not be null, checking in listener
 	 */
-	void handleConsentUpdate(final Event event) {
-		processAddEvent(event);
+	void handleConsentUpdate(@NonNull final Event event) {
+		processEdgeEvent(event);
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:243]: a bit confusing cause this is not an edge event - might be good to rename it to processEvent

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java [hunk 11]
```diff
@@ -573,33 +387,14 @@ Collection<Event> getCachedEvents() {
...
-		if (configSharedState == null) {
-			MobileCore.log(
-				LoggingMode.DEBUG,
-				LOG_TAG,
-				"EdgeExtension - Configuration is pending, couldn't process event at this time, waiting..."
-			);
+		SharedStateResult sharedStateResult = getApi()
+			.getSharedState(EdgeConstants.SharedState.CONFIGURATION, event, false, SharedStateResolution.ANY);
+		if (sharedStateResult == null || sharedStateResult.getStatus() != SharedStateStatus.SET) {
+			Log.debug(LOG_TAG, LOG_SOURCE, "Configuration is pending, couldn't process event at this time, waiting...");
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:393]: this should not be the case as we check in ready for event, but this message is now confusing as there is no more waiting at this time since the event was already processed and there is no retry after that. I suggest we raise this to a warning mentioning this is unexpected and remove the waiting part

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java [hunk 12]
```diff
@@ -609,32 +404,18 @@ public void error(ExtensionError extensionError) {
...
-			MobileCore.log(
-				LoggingMode.DEBUG,
+		SharedStateResult sharedStateResult = getApi()
+			.getXDMSharedState(EdgeConstants.SharedState.IDENTITY, event, false, SharedStateResolution.ANY);
+		if (sharedStateResult == null || sharedStateResult.getStatus() != SharedStateStatus.SET) {
+			Log.debug(
 				LOG_TAG,
-				"EdgeExtension - Identity shared state is pending, could not process queued events at this time, waiting..."
+				LOG_SOURCE,
+				"Identity shared state is pending, could not process queued events at this time, waiting..."
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:413]: same comment here

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java [hunk 13]
```diff
@@ -609,32 +404,18 @@ public void error(ExtensionError extensionError) {
...
-					}
-				}
-			);
-
-		if (identityXDMSharedState == null) {
-			MobileCore.log(
-				LoggingMode.DEBUG,
+		SharedStateResult sharedStateResult = getApi()
+			.getXDMSharedState(EdgeConstants.SharedState.IDENTITY, event, false, SharedStateResolution.ANY);
+		if (sharedStateResult == null || sharedStateResult.getStatus() != SharedStateStatus.SET) {
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:409]: do we need to check the status again since it was verified in readyForEvent or can we go ahead and fetch the value directly

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeState.java (hunk 9 above)
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeState.java:69]: I'm not sure why I said Identity here. The hub's state is needed to build ImplementationDetails using the Core version and wrapper type, and setting the default Consent if Consent is not registered. Both of those a checked in the synchronized block, in `ImplementationDetails.fromEventHubState` and `updateCollectConsentIfNotRegistered`.
I'll update the comment to make it more clear.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java [hunk 14]
```diff
@@ -81,279 +87,137 @@ protected EdgeExtension(final ExtensionApi extensionApi) {
...
+		if (EventUtils.isExperienceEvent(event) || EventUtils.isUpdateConsentEvent(event)) {
+			SharedStateResult configurationState = getApi()
+				.getSharedState(EdgeConstants.SharedState.CONFIGURATION, event, false, SharedStateResolution.ANY);
+			SharedStateResult identityState = getApi()
+				.getXDMSharedState(EdgeConstants.SharedState.IDENTITY, event, false, SharedStateResolution.ANY);
+			return (
+				configurationState != null &&
+				identityState != null &&
+				configurationState.getStatus() == SharedStateStatus.SET &&
+				identityState.getStatus() == SharedStateStatus.SET
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:204]: I can reorder them but the formatter still places each on a new line.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java (hunk 10 above)
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:243]: These are events which go to the Edge Network. I want to differentiate the "events" here from all the other events processed in this class, so I'd like something less generic than processEvent. What about `processEventToHitQueue` or `sendEventToHitQueue` or `queueEdgeDataEvent` ?

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java (hunk 11 above)
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:393]: I reorganized this a little to make it cleaner. I moved the warning messages to the (soon to be renamed) processEdgeEvent method (same as in iOS) and then I'm using these getter in readyForEvent as well so the code to retrieve the shared states is the same both when it is checked (readyForEvent) and when received (processEdgeEvent).

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java (hunk 13 above)
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:409]: It's safer to check and I don't think it harms performance in any meaningful way. In iOS we nil check the shared state as well, it's just that the Swift language semantics makes it a one liner.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java]: I refactored this code so these checks are now moved out of readyForEvent.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeConstants.java [hunk 15]
```diff
@@ -14,7 +14,7 @@
...
 	static final String EXTENSION_NAME = "com.adobe.edge";
 	static final String FRIENDLY_NAME = "Edge";
 	static final String LOG_TAG = FRIENDLY_NAME;
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeConstants.java:20]: can we now remove EventSource and EventType or will you do that in the next PR?

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java (hunk 10 above)
Comment (emdobrin) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:225]: or `processAndQueueEvent`

**Review by emdobrin — COMMENTED** [review]
Changes look good to me with couple small follow-ups

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeConstants.java (hunk 15 above)
Comment (kevinlind) [reply code/edge/src/main/java/com/adobe/marketing/mobile/EdgeConstants.java:20]: Yes, in another PR. It's actually part 3 of the 2.0 adoption.

File: code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java [hunk 16]
```diff
@@ -309,9 +309,9 @@ void handleSetLocationHint(@NonNull final Event event) {
 	}
 
 	/**
-	 * Processes an Experience Event or Consent Update Event.
+	 * Processes an Experience Event or Consent Update Event and add to the hit queue.
```
Comment (emdobrin) [inline code/edge/src/main/java/com/adobe/marketing/mobile/EdgeExtension.java:312]: `... adds it to the hit queue`

**Review by emdobrin — APPROVED** [review]
